Adds a dotnet test --logger to invoke Formatter

clrudolphi opened the pull request:
### 🤔 What's changed?

This PR adds a custom vstest infrastructure logger that can be used via the command line to configure Formatters.
It will pick up and use the TestResults folder if one is not explicitly specified.

ex: 
```powershell
dotnet test --logger:"formatter-message;outputFilePath=foo.ndjson"
``` 

### ⚡️ What's your motivation? 

This does get us access to the TestResults path which otherwise isn't easily available.

### 🏷️ What kind of change is this?

- :zap: New feature (non-breaking change which adds new behaviour)


### ♻️ Anything particular you want feedback on?

Anything.

### 📋 Checklist:


- [X] I've changed the behaviour of the code
  - [x] I have added/updated tests to cover my changes.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "[vNext]" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request & included my GitHub handle to the release contributors list.

----

*This text was originally taken from the [template of the Cucumber project](https://github.com/cucumber/.github/blob/main/.github/PULL_REQUEST_TEMPLATE.md), then edited by hand. [You can modify the template here.](https://github.com/reqnroll/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

clrudolphi committed:
Initial commit.

clrudolphi committed:
First working example. Takes command line input as a --logger and converts the input to an override to be applied via Environment Variable.

gasparnagy committed:
Merge branch 'main' into TestLogger

clrudolphi commented:
Agree on all your points. I like the first option. 

As an aside, another use for a custom logger would be for us internally. We often struggle to log the internal operation of Reqnroll and to keep such log content separate from the content of the test execution. If we generalized the concept of the IFormatterLog and sent that content only to a custom --Logger, that would provide us a convenient way to log internal execution state.

clrudolphi commented:
Another thought- I agree that the logger should set a distinct environment variable. 
Should that value overwrite the configuration for the respective formatter OR be in addition to it? 

In other words we might have one message sink configured via json file and another configured via command line? I could see a configuration provider that watches for the special environment variable and then adds a new instance of the formatter to the global container. As long as this happens early enough in the startup sequence it would work.

gasparnagy commented:
> Another thought- I agree that the logger should set a distinct environment variable. Should that value overwrite the configuration for the respective formatter OR be in addition to it?
> 
> In other words we might have one message sink configured via json file and another configured via command line? I could see a configuration provider that watches for the special environment variable and then adds a new instance of the formatter to the global container. As long as this happens early enough in the startup sequence it would work.

I think this should be additive/overwrite:

1. If I configured html from JSON, but use --logger with message: it should enable both
2. If I configured html from JSON to save file to x.html, but use --logger with file y.html: it should keep html enabled, but overwrite the output to y.html.

The interesting situation is, what should be the result, if I have x.html configured in JSON (that means bin/Debug/net9.0/x.html), but would "enable" is also from the logger, but without an output file, then it should be
1. Keep it bin/Debug/net9.0/x.html
2. Use the output folder from logger: TestResults/x.html
3. Use the default output file from logger: TestResults/reqnroll_report.html

I think we should go for 1). That is the easiest anyway, and the use-case is pretty much artificial.

clrudolphi committed:
Modified per reveiw comments. The format string for the command line now uses key-value pairs, such as --logger "formatter;name=message;outputFilePath=foo.ndjson". The infrastructure is now in place to set multiple loggers on the command line, but that is not  yet working.

clrudolphi committed:
Fixed tests to be OS-path aware.

clrudolphi committed:
Created two Formatter Loggers (one for each built-in formatter) that derive from a base class.

clrudolphi commented:
As the first portion of refactoring this, commit f99c7d3 creates two named loggers (one for messages and one for html).
Usage:
`--logger "formatter-html;outputFilePath=foo.html" --logger "formatter-message;outputFilePath=foo.html"`

If you prefer the names reversed (such as `html-formatter`) let me know.
These support configuring the file path in KVP form.

I'll next add another more generic formatter logger (will call it `formatters` {plural}) and require the configuration string to be encoded json. This form will be able to support any combination of built-in and user-plug-in formatters.

clrudolphi commented:
@gasparnagy I've tried to create a  --logger that would accept a general json config element and I'm failing. I can't get the quote escaping correct. Powershell and/or the dotnet test infrastructure seem to strip out all quotes from the command line parameter and I'm left with a string that I can't pass to the Formatter configuration subsystem. I don't have confidence that I could parse this string and reconstruct a proper json representation of it (as its supposed to be general, we can't predict what other plugin formatters may want as configuration input).

So far, I've tried:

- surrounding the string with single-quotes
- escaping quotes with back-ticks: 
```
--logger 'formatters;config={`"formatters`":{`"message`":{`"outputFilePath`":`"foo.ndjson`"}}}'
```
- doubling up the double-quotes (as in, `--logger "formatters;config={""formatters"":{""message"":{""outputFilePath"":""foo.ndjson""}}}"`
- using powershell variables
```
 $config = '{"formatters":{"message":{"outputFilePath":"foo.ndjson"}}}'
PS C:\Users\clrud\source\repos\scratch\vs95\vs95> $config
{"formatters":{"message":{"outputFilePath":"foo.ndjson"}}}
PS C:\Users\clrud\source\repos\scratch\vs95\vs95> dotnet test -c Debug --logger "formatters;config=$config" 
```

I'm open to suggestions of things to try. Any ideas?

clrudolphi committed:
Corrected encoding of nuspec file

gasparnagy commented:
@clrudolphi sorry for not responding earlier.

> As the first portion of refactoring this, commit [f99c7d3](https://github.com/reqnroll/Reqnroll/commit/f99c7d3bbf05206c9d20e425b5287b42ebf0b34d) creates two named loggers (one for messages and one for html). Usage: `--logger "formatter-html;outputFilePath=foo.html" --logger "formatter-message;outputFilePath=foo.html"`

This is great! I like it. 

> If you prefer the names reversed (such as `html-formatter`) let me know. These support configuring the file path in KVP form.

I see the problem. Not easy. Probable the reverse name would be better though.

> I've tried to create a --logger that would accept a general json config element and I'm failing.

I have also tried that once (other project), but running into the same issues. I think this is not feasible, because others will suffer from the same problems. 

My suggestion is the following:

The generic one should be called "formatter" (singular) and should only allow configuring a single formatter, in a form `--logger "formatter;name=message;outputFilePath=foo.ndjson"`. As we have specific one for html and message, it is unlikely that someone want to configure two custom formatters from the command line, but if so, they can still do it via environment variables.

clrudolphi committed:
Renamed the built-in loggers to html-formatter and message-formatter

clrudolphi committed:
Corrrecting file encoding (again - sorry)

clrudolphi commented:
Before we merge this, I'll edit the docs to include the logger configuration of the formatters.
@gasparnagy And before that, would you mind reviewing the PR with documentation edits that I made earlier today? Those should probably be merged before I edit the doc again.

clrudolphi committed:
Documentation

Added description of how to configure the formatters using the --logger command line.

obligaron commented on Reqnroll.MessagesLogger.TestLogger/FormatterTestLogger.csproj:
```diff
@@ -0,0 +1,17 @@
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <AssemblyName>FormatterTestLogger</AssemblyName>
+    <PackageId>FormatterTestLogger</PackageId>
```
Should both start with `Reqnroll.`?

clrudolphi committed:
Cleanup of Reqnroll.csproj project references

Co-authored-by: obligaron <obligaron@live.com>

clrudolphi committed:
Changed name of Logger project Reqnroll.FormatterTestLogger.

gasparnagy commented on Reqnroll.MessagesLogger.TestLogger/FormatterTestLogger.csproj:
```diff
@@ -0,0 +1,17 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="17.14.1" />
```
Hm. I have checked the linked article, and based on that we should strongly consider if we want to have this logger support included. It is not essential for us and it is declared to be slowly deprecated.

Later we can think about having an MTP reporter.

gasparnagy committed:
Merge remote-tracking branch 'origin/main' into TestLogger

gasparnagy committed:
code cleanup

gasparnagy committed:
more code cleanup

gasparnagy committed:
Cleanup Reqnroll.FormatterTestLogger

gasparnagy committed:
code cleanup

gasparnagy committed:
replace GetEnvironmentVariableNames with GetEnvironmentVariables
